### ISSUE TITLE
Association query in Doctrine2 cannot be properly built

### ISSUE THREAD

**Issue by alexkunin** [Codeception/Codeception#5542]
This is related to #4169 -- probably can be considered a duplicate, but I provide complete steps to reproduce, analysis, and solution. So, I hope it deserves separate issue.

#### What are you trying to achieve?

I'm trying to use `canSeeInRepository` to check if specific entity and some other associated entities are in database, something like this:

```php
$I->seeInRepository(\App\Entity\User::class, [
    'name' => 'User1',
    'projects' => [
        'name' => 'Project1',
        'status' => [
            'name' => 'Status1',
        ],  
    ],  
]); 
```

Expected result would be a successful or failed test.

#### What do you get instead?

I get an exception from depths of Doctrine2 module, something like this:

```
1) FirstCest: Try to test
 Test  tests/functional/FirstCest.php:tryToTest
                                                                                                                                                                                                                                                                                                                                       
  [TypeError] Argument 1 passed to Doctrine\Common\Persistence\Mapping\
AbstractClassMetadataFactory::getRealClass() must be of the type string
 null given, called in <...>/vendor/doctrine/persistence/lib/Doctrine/Common/
Persistence/Mapping/AbstractClassMetadataFactory.php on line 165  
```

### Details

* Codeception version: 3.0.1
* PHP Version: 7.3.3
* Operating System: macOS
* Installation type: Composer

### What is going on inside Doctrine2 module?

There is the following method:

```php
protected function buildAssociationQuery($qb, $assoc, $alias, $params)
{
    $data = $this->em->getClassMetadata($assoc);
    foreach ($params as $key => $val) {
        if (isset($data->associationMappings)) {
            if ($map = array_key_exists($key, $data->associationMappings)) {
                if (is_array($val)) {
                    $qb->innerJoin("$alias.$key", "_$key");
                    foreach ($val as $column => $v) {
                        if (is_array($v)) {
                            $this->buildAssociationQuery($qb, $map['targetEntity'], $column, $v);
                            continue;
                        }
                        $paramname = "_$key" . '__' . $column;
                        $qb->andWhere("_$key.$column = :$paramname");
                        $qb->setParameter($paramname, $v);
                    }
                    continue;
                }
            }
        }
        if ($val === null) {
            $qb->andWhere("s.$key IS NULL");
        } else {
            $paramname = str_replace(".", "", "s_$key");
            $qb->andWhere("s.$key = :$paramname");
            $qb->setParameter($paramname, $val);
        }
    }
}
```

The method tries to build a query according to input array, but fails in few places:

1. `$map` is meant to get association description from `$data->associationMappings`, but in reality it gets `true` value.
2. If association sub-query is found in input array, it adds a JOIN to query and calls itself recursively, but does this incorrectly: it passes wrong alias to recursive call, and also does a separate recursive call for each field in associated record.
3. JOIN query from p.2 uses association name as alias, which can cause issues if there are several associations with the same name (e.g. every entity could have `owner` association).
4. In all recursive calls it uses hardcoded alias "s" instead of one from arguments, which results in invalid fields being checked against passed values. 

Here is fixed version:

```php
protected function buildAssociationQuery($qb, $assoc, $alias, $params)
{
    $data = $this->em->getClassMetadata($assoc);
    foreach ($params as $key => $val) {
        if (isset($data->associationMappings)) {
            if (array_key_exists($key, $data->associationMappings)) {
                $map = $data->associationMappings[$key];
                if (is_array($val)) {
                    $qb->innerJoin("$alias.$key", "${alias}__$key");
                    $this->buildAssociationQuery($qb, $map['targetEntity'], "${alias}__$key", $val);
                    continue;
                }
            }
        }
        if ($val === null) {
            $qb->andWhere("$alias.$key IS NULL");
        } else {
            $paramname = str_replace(".", "", "${alias}_$key");
            $qb->andWhere("$alias.$key = :$paramname");
            $qb->setParameter($paramname, $val);
        }
    }
}
```

It fixes all mentioned problems, and as you can see in "Steps to reproduce", the exception is gone. Here is an example of produced DQL (you'll get exactly this DQL if you follow steps to reproduce; added formatting for clarity):

```sql
SELECT
    s
FROM
    App\Entity\User s
    INNER JOIN s.projects s__projects
    INNER JOIN s__projects.status s__projects__status
WHERE
    s.name = :s_name
    AND s__projects.name = :s__projects_name
    AND s__projects__status.name = :s__projects__status_name
```

I'll try to add a pull request with the fix later.

### Steps to reproduce:

Here are complete steps to reproduce the issue in a new Symfony project. Symfony is not relevant here, it just allows for quick setup of the environment (specifically Doctrine part).

```bash
# make sure PHP interpreter and composer are available

# create new Symfony project:
composer create-project symfony/skeleton Codeception-Doctrine2-case
cd Codeception-Doctrine2-case

# add few requirements:
composer require orm
composer require symfony/maker-bundle codeception/codeception --dev
# you'll need to confirm codeception bootstrap script execution

# enable Doctrine2 module by uncommenting related lines in config file:
vim tests/functional.suite.yml 

# change DATABASE_URL in .env to "sqlite:///%kernel.project_dir%/var/app.db":
vim .env

# initialize database:
bin/console doctrine:database:create

# make Status entity:
bin/console make:entity
# use "Status" as class name
# enter "name" as first field name and press Enter 3 times to accept defaults:
#   first Enter to confirm default field type "string"
#   second Enter to confirm default field length "255"
#   third Enter to confirm the field is non-nullable
# then press Enter to avoid adding fields

# make Project entity:
bin/console make:entity
# use "Project" as class name
# enter "name" as first field name and press Enter 3 times to accept defaults
# use "status" as field second name
# use "ManyToOne" as field type
# Use "Status" as related class
# then press Enter several (4?) times to confirm defaults and stop adding new fields

# make User entity:
bin/console make:entity
# use "User" as class name
# enter "name" as first field name and press Enter 3 times to accept defaults
# use "projects" as field second name
# use "OneToMany" as field type
# Use "Project" as related class
# then press Enter several (3?) times to confirm defaults and stop adding new fields

# create migration and update database:
bin/console make:migration
bin/console doctrine:migrations:migrate

# create new functional test:
vendor/bin/codecept g:cest functional First

# open test file and add following code to "tryToTest" method:
vim tests/functional/FirstCest.php
$I->seeInRepository(\App\Entity\User::class, [
    'name' => 'User1',
    'projects' => [
        'name' => 'Project1',
        'status' => [
            'name' => 'Status1',
        ],  
    ],  
]); 

# run tests:
vendor/bin/codecept run
# you'll see the exception mentioned in issue 

# edit composer.json to use fixed fork of codeception:
vim composer.json
# add "@dev" to codeception version, so the section looks like this:
    "require-dev": {
        "codeception/codeception": "^3.0@dev",
        "symfony/maker-bundle": "^1.11"
    }
# add new "repositories" section at the end:
    "repositories": [
        {
            "type": "git",
            "url": "https://github.com/alexkunin/codeception"
        }
    ]
# then reinstall codeception:
composer update

# run tests again, now with "--debug" to see DQL; this time there will be failed test, not an exception:
vendor/bin/codecept run --debug
# and in the output there will be correct DQL
```

**Comment by alexkunin** [comment]
I have added the pull request, but it does not contain any tests. Doctrine2 module has no tests whatsoever, and I don't even known how write one for a complex project like this. Per issue template, tests are welcome but not required, so, I hope this small fix can be uses as is, without tests.

As for docs, the fix does not change any behavior (as far as I can tell), and info in docs is still relevant: https://codeception.com/docs/modules/Doctrine2#seeInRepository. It could benefit from some more examples, but thats a bit different story which is unrelated to this issue.

**Comment by Naktibalda** [comment]
There are some tests that are using Doctrine2 module in codeception/symfony-demo repo,
they are executed as part of Codeception travis build.

https://github.com/Codeception/symfony-demo/blob/master/tests/functional/PostCrudCest.php

**Comment by alexkunin** [comment]
@Naktibalda, that really helps, thank you.

But now I'm facing some questions:

1. Should I continue discussing this here (main repo) or in `symfony-demo`? Most of the issue is related directly to `Doctrine2` module of this repo, and the other repo's goal is not to test functionality in question -- it just happens to partially cover the code we're talking about.

2. That repo specifies very strict version of `codeception`: "2.5.x-dev" (verbatim copy from `composer.json`). My fix can be applied there with a separate pull request (the code in question was not changed for 2 years old, last update was in 2.2). But what about version 3.0? Updating `symfony-demo` to use 3.0 is well beyond my capabilities. What I can do here?

3. That repo has quite specific purpose: to demo Codeception as a tool for Symfony developers, so, the tests kinda touch Doctrine specifics, but very tangentially. I have identified suitable test which I can modify to showcase the issue (`CommentCest`), but resulting test would really look like a synthetic one:

```php
// before:
$I->seeInRepository(Comment::class, ['content' => 'Hi, Symfony!']);

// after:
$I->seeInRepository(Comment::class, [
    'content' => 'Hi, Symfony!',
    'author' => [
        'username' => 'john_user',
    ],  
    'post' => [
        'comments' => [
            'content' => 'Hi, Symfony!',
            'author' => [
                'username' => 'john_user'
            ]   
        ]   
    ]   
]); 
```

As you can see, existing test just checks if new comment is in database. New version checks:

* is new comment in DB?
* does its author has correct username?
* does this comment belong to post which has a comment with correct content and author?

Well, it does the job, but this is really far from optimal testing practices (as far as I can tell). And I don't think I can come up with a better test without heavily altering `symfony-demo` (which, again, is beyond my capabilities). But again, it does the job (fails with old code, passes with the fix applied), so maybe it is acceptable to have such a synthetic assertion in a demo repo if we add some relevant comment?

4. Docs. Previously I stated that they are OK, and that no update is required, but now I'm not so sure about that. Here is a sample snippet from docs:

```php
$I->seeInRepository('AppBundle:User', array('name' => 'davert'));
$I->seeInRepository('User', array('name' => 'davert', 'Company' => array('name' => 'Codegyre')));
$I->seeInRepository('Client', array('User' => array('Company' => array('name' => 'Codegyre')));
```

It seems that association is meant to be identified by target class (array key is `User` which matches entity class name, and also `Company`), not by association name (which I'd expect to be `user` and `company`). But in the code I see that keys are checked against association names, not against entity class names. So, example in docs is misleading, and probably needs fixing?

**Comment by Naktibalda** [comment]
1. Keep it here.
2. It works with Codeception 3.0, when checkout by Codeception build process: https://travis-ci.org/Codeception/Codeception/jobs/543654420
composer.json only applies to the builds of symfony-demo build, so update that constraint as part of your change.
3. It's purpose isn't demo, it is named symfony-demo, because it is a clone of symfony/demo repo. But it serves as a testsuite of symfony integration.
If that test demonstrates the issue, it is good enough. Add a link to this issue to the docblock.
4. If the docs are misleading, please update them. @DavertMik any comments about that example?

**Comment by alexkunin** [comment]
@Naktibalda, thank you for detailed response!

2. Yep, I see that now, thanks. Had to switch to PHP 7.2, though, otherwise was observing same errors as in #5523. Also, that repo states in README.md that one should use `vendor/bin/simple-phpunit` to run tests, which is a bit misleading.

Anyhow, after fixing PHP version and injecting my branch of Codeception, new test assertion passed.

3. OK, I see, and that's probably why one should not try to fix README.md.

So, I have forked that repo, bumped codeception dependency, modified assertion, and added detailed comment next to it. It fails with current Codeception repo, but succeeds if I use my fix (with "repositories" hack described above).

Pull request was created (https://github.com/Codeception/symfony-demo/pull/16). And, as one could expect, it fails due to the exception described in the opening post in this issue. I guess it will become green when my fix to Codeception repo is merged. Looks kinda like chicken and egg problem for me, not sure how to handle it properly.

4. I did not update docs for now, it's a tricky topic for me. Let's wait what for @DavertMik response.
